### DISCUSSION TITLE
코드 컨벤션 어떻게 할까요?

### DISCUSSION THREAD

**Discussion by laigasus** [Strong-Potato/TripVote-BE#10] [General]
할 것들을 나열해봤습니다.
정해진 내용은 체크 박스 활성화해주세요. 추후 Wiki 에 올리겠습니다!

- [x] 공통 예외 처리
- [x] 응답 Api Response
- [x] Git 전략, 커밋 컨벤션, 템플릿 정형화

**Comment by happymink** [comment]
커밋 컨벤션이 정해지면 인텔리제이 Commit Message Template 플러그인을 사용하면 유용할 것 같기도 합니다!

참고링크 : https://zzang9ha.tistory.com/334#%F0%9F%93%8E%C2%A0-commit-message-template

**Reply by NoJaeHyuk** [reply]
![image](https://github.com/Strong-Potato/TripVote-BE/assets/32382839/35f10624-116a-42ec-9fbd-211b54b4a19f)

**Reply by NoJaeHyuk** [reply]
Git Commit Template

**Reply by NoJaeHyuk** [reply]
![image](https://github.com/Strong-Potato/TripVote-BE/assets/32382839/ed5d3bbf-b348-4224-bcec-d0e4468cabf4)

**Reply by NoJaeHyuk** [reply]
https://plugins.jetbrains.com/plugin/9861-git-commit-template

**Comment by laigasus** [comment]
예외처리는 BIz, 공통단 별로 `exception` 패키지를 사용했습니다.
Biz에서 예외를 만들는데 파일이 꽤 많아지더군요
저는 개인적으로 springboot 3.x 에서 지원하는 [rfc 7807](https://datatracker.ietf.org/doc/html/rfc7807) ProblemDetails 를 이용해서 가공했습니다!

작업한 사례도 올립니다.
https://github.com/Mini-Team-6/Mini-Team-6-Backend

**Reply by laigasus** [reply]
요런 느낌
![image](https://github.com/Strong-Potato/TripVote-BE/assets/26517061/ce397eb4-8859-40dc-ba8b-8af95b8460c5)

**Reply by Dr-KoKo** [reply]
![image](https://github.com/Strong-Potato/TripVote-BE/assets/97681286/92b8eea8-8133-4262-8dc7-c2eae8ca8ee0)
`Throwable`은 테스트 용도로 사용하던 거라서 제외하고 보시면 될 것 같습니다.

**Reply by NoJaeHyuk** [reply]
![image](https://github.com/Strong-Potato/TripVote-BE/assets/32382839/b5d6c756-4b34-43f0-909f-14d3e64115af)

상태코드와 메세지만 받는 형식으로 주로 사용했습니다.

**Comment by NoJaeHyuk** [comment]
```
    @PostMapping("/register")
    public ResponseEntity<ApiResponse<Void>> register(@Valid @RequestBody MemberRegisterRequest request) {
        memberRegisterService.register(request);
        return ResponseEntity.ok().body(new ApiResponse<>(LocalDateTime.now(), "성공", null));
    }

    @GetMapping("/email/verify")
    public ResponseEntity<ApiResponse<Void>> verifyEmail(@NotNull String email) {
        memberRegisterService.validateEmailDuplicatation(email);
        return ResponseEntity.ok().body(new ApiResponse<>(LocalDateTime.now(), "성공", null));
    }


    //// 

    public record ApiResponse<T>(
    LocalDateTime timeStamp,
    String detail,
    T data) {

    }
```

**Comment by NoJaeHyuk** [comment]
`
@Getter
public class ApiResponse<T> {

    private int code;
    private HttpStatus status;
    private String message;
    private T data;

    public ApiResponse(HttpStatus status, String message, T data) {
        this.code = status.value();
        this.status = status;
        this.message = message;
        this.data = data;
    }

    public static <T> ApiResponse<T> of(HttpStatus status, String message, T data) {
        return new ApiResponse(status, message, data);
    }

    public static <T> ApiResponse<T> of(HttpStatus status, T data) {
        return of(status, status.name(), data);
    }

    public static <T> ApiResponse<T> ok(T data) {
        return of(HttpStatus.OK, HttpStatus.OK.name(), data);
    }
}
`

**Comment by laigasus** [comment]
작업이 모두 완료되었음 https://github.com/Strong-Potato/TripVote-BE/pull/18
